### PR TITLE
Lidar parsing

### PR DESCRIPTION
**Pull request by Mmoyv27**
Added the lidar_parser module

### CODE REVIEW

File: modules/lidar_parser/lidar_parser_worker.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+"""
+Gets detections and outputs oscillations when complete.
+"""
+
+from modules import lidar_detection
+from modules import lidar_oscillation
+from modules import decision
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser_worker.py]: remove decision import - not needed here

File: modules/lidar_parser/lidar_parser.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+"""
+Module to parse LiDAR data, detect oscillations, and return oscillation objects.
+"""
+
+from modules import lidar_oscillation
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser.py]: small nit - put this in alphabetical order

File: modules/lidar_parser/lidar_parser.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+class LidarParser:
+    """
+    Class to handle parsing of LiDAR data stream and detecting complete oscillations.
+    """
+
+    __create_key = object()
+
+    @classmethod
+    def create(cls) -> "tuple[bool, LidarParser | None]":
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser.py]: no need for a create method for this module

File: modules/lidar_parser/lidar_parser.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        else:
+            print("Error creating LidarOscillation")
+
+        self.lidar_readings = []
+
+    def run(self, lidar_detection: lidar_detection.LidarDetection) -> None:
+        """
+        Process a single LidarDetection.
+        """
+        if not self.__run:
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser.py]: why do we need this check? the run function is called in a loop not just once

File: modules/lidar_parser/lidar_parser.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        """
+        Private constructor for LidarParser. Use create() method.
+        """
+        assert class_private_create_key is LidarParser.__create_key, "Use the create() method"
+
+        self.lidar_readings = []
+        self.current_oscillation = None
+
+        self.last_angle = None
+        self.direction = None
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser.py]: let's create an enum class for the direction of the lidar scan

File: modules/lidar_parser/lidar_parser_worker.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+def lidar_oscillation_worker(
+    detection_in_queue: queue_wrapper.QueueWrapper,
+    oscillation_out_queue: queue_wrapper.QueueWrapper,
+    controller: worker_controller.WorkerController,
+) -> None:
+    """
+    Feeding LidarParser continuously with a stream of LidarDetection
+    """
+
+    result, parser = lidar_parser.LidarParser.create()
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser_worker.py]: no create method

File: modules/lidar_parser/lidar_parser_worker.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        return
+
+    while not controller.is_exit_requested():
+        controller.check_pause()
+
+        lidar_data: lidar_detection.LidarDetection = detection_in_queue.queue.get()
+        if lidar_data is None:
+            break
+
+        parser.run(lidar_data)
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser_worker.py]: this should return a result and either none or a lidar osciallation

File: modules/lidar_parser/lidar_parser_worker.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+        result, oscillation = parser.run(detection_in_queue)
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser_worker.py]: the run function accepts a LidarDetection as a parameter. so we don't pass the queue to the run function.

**Review by ashum68 — COMMENTED** [review]
Reviewed. 

I think we can make an optimization to how we return a complete oscillation. instead of creating a LidarOscillation in a separate function and assigning it to a variable (extra memory) we can directly return it inside of process_reading(). i.e. `if ... : return lidar_oscillation.LidarOscillation.create(self.lidar_readings)` and then reinitialize lidar_readings

Also, it seems like run is just a wrapper for process reading so we can just move the contents of process_reading into the run function

File: modules/lidar_parser/lidar_parser_worker.py (hunk 1 above)
Comment (ashum68) [reply modules/lidar_parser/lidar_parser_worker.py]: can you also refer to https://uwarg-docs.atlassian.net/wiki/spaces/CV/pages/2253226033/Python+Style+Convention#Imports for import order

File: modules/lidar_parser/lidar_parser_worker.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+def lidar_oscillation_worker(
+    detection_in_queue: queue_wrapper.QueueWrapper,
+    oscillation_out_queue: queue_wrapper.QueueWrapper,
+    controller: worker_controller.WorkerController,
+) -> None:
+    """
+    Feeding LidarParser continuously with a stream of LidarDetection
+    """
+
+    result, parser = lidar_parser.LidarParser()
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser_worker.py]: this doesn't return a result - just the parser

File: modules/lidar_parser/lidar_parser_worker.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    if not result:
+        print("Failed to initialise LidarParser.")
+        return
+
+    while not controller.is_exit_requested():
+        controller.check_pause()
+
+        lidar_reading: lidar_detection.LidarDetection = detection_in_queue.queue.get()
+        if lidar_reading is None:  # Do we want this line?
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser_worker.py]: remove this comment

File: modules/lidar_parser/lidar_parser.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        self.direction = Direction.NONE
+
+    def run(
+        self, lidar_detection: lidar_detection.LidarDetection
+    ) -> "tuple[bool, lidar_oscillation.LidarOscillation | None]":
+        """
+        Process a single LidarDetection and return the oscillation if complete.
+        """
+
+        if lidar_detection is None:
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser.py]: worker ensures lidar_detection is not None

File: modules/lidar_parser/lidar_parser_worker.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+from worker import worker_controller
+
+
+def lidar_oscillation_worker(
+    detection_in_queue: queue_wrapper.QueueWrapper,
+    oscillation_out_queue: queue_wrapper.QueueWrapper,
+    controller: worker_controller.WorkerController,
+) -> None:
+    """
+    Feeding LidarParser continuously with a stream of LidarDetection
```
Comment (ashum68) [inline modules/lidar_parser/lidar_parser_worker.py]: small nit -  add a period

File: tests/unit/test_lidar_parser.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
+"""
+Unit tests for lidar_parser module.
+"""
+
+import pytest
+
+from modules import lidar_detection
+from modules.lidar_parser import lidar_parser
+
+ANGLE_UP = 15.0
```
Comment (ashum68) [inline tests/unit/test_lidar_parser.py]: these can just be hard-coded

File: tests/unit/test_lidar_parser.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+        higher_angle_detection_up: lidar_detection.LidarDetection,
+        lidar_detection_down: lidar_detection.LidarDetection,
+    ) -> None:
+        """
+        Test detection of an oscillation with a change in direction from up to down after multiple detections.
+        """
+        lidar_parser_instance.run(lidar_detection_up)
+        lidar_parser_instance.run(higher_angle_detection_up)
+        result, oscillation = lidar_parser_instance.run(lidar_detection_down)
+        assert result
```
Comment (ashum68) [inline tests/unit/test_lidar_parser.py:96]: check list length (i.e. len == 2) and check whether the oscillation contains only the readings going up i.e. it sees 15 and 20 degrees. (lidar_oscillation.readings[0].angle == 15)

File: tests/unit/test_lidar_parser.py [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+    ) -> None:
+        """
+        Test that the parser resets its readings after detecting an oscillation.
+        """
+        lidar_parser_instance.run(lidar_detection_up)
+        lidar_parser_instance.run(higher_angle_detection_up)
+        result, oscillation = lidar_parser_instance.run(lidar_detection_down)
+        assert result
+        assert oscillation is not None
+        assert len(lidar_parser_instance.lidar_readings) == 0
```
Comment (ashum68) [inline tests/unit/test_lidar_parser.py]: store last reading after direction switch

File: tests/unit/test_lidar_parser.py [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+        lidar_parser_instance.run(higher_angle_detection_up)
+        result, oscillation = lidar_parser_instance.run(lidar_detection_down)
+
+        if result:
+            oscillation_count += 1
+            assert oscillation is not None
+
+        for i in range(4):
+            if i % 2 == 0:
+                result, oscillation = lidar_parser_instance.run(higher_angle_detection_up)
```
Comment (ashum68) [inline tests/unit/test_lidar_parser.py:168]: check len

File: tests/unit/test_lidar_parser.py (hunk 14 above)
Comment (ashum68) [reply tests/unit/test_lidar_parser.py:96]: 1. expected = lidar_detection.create(5, 10)
2. oscillation = parser.run()
3. assert expected.angle == oscillation.readings[0].angle

Comment (Mmoyv27) [reply tests/unit/test_lidar_parser.py:96]: I've implemented this in the test_oscillation_detected_up_down

File: tests/unit/test_lidar_parser.py [hunk 17]
```diff
@@ -188,6 +187,7 @@ def test_no_oscillation_with_single_reading(
         result, oscillation = lidar_parser_instance.run(lidar_detection_up)
         assert not result
         assert oscillation is None
+        assert lidar_parser_instance.lidar_readings[0].angle == 15.0
```
Comment (ashum68) [inline tests/unit/test_lidar_parser.py:190]: we want to check that the returned oscillation has the angle 15 in it not the instance which is the test case that you created.

File: tests/unit/test_lidar_parser.py [hunk 18]
```diff
@@ -102,6 +99,8 @@ def test_oscillation_detected_up_to_down(
         result, oscillation = lidar_parser_instance.run(lidar_detection_down)
         assert result
         assert oscillation is not None
+        assert len(oscillation.readings) == 2
+        assert lidar_parser_instance.lidar_readings[0].angle == -15.0
```
Comment (ashum68) [inline tests/unit/test_lidar_parser.py]: check that the oscillation has this angle in it, not the instance that you created

**Review by ashum68 — COMMENTED** [review]
reviewed

File: tests/unit/test_lidar_parser.py (hunk 17 above)
Comment (Mmoyv27) [reply tests/unit/test_lidar_parser.py:190]: oscillation not being returned here, as solo angle being provided to our lidar_parser instance. so the last assert basically looks at the leftovers, that the 15 degree ange is being processed, but not returned

**Review by ashum68 — APPROVED** [review]
LGTM

**Pull request merged**